### PR TITLE
Divide the circle : find the indices, number of pixels, pixel indices 

### PR DESCRIPTION
**Pull request by sameera2004**

### CODE REVIEW

File: skxray/diff_roi_choice.py [hunk 1]
```diff
@@ -375,3 +378,117 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
...
+
+        mesh[vl] = i + 1
+
+    roi_inds, num_pixels, pixel_list = _process_rois(np.ravel(mesh),
+                                                     detector_size,
+                                                     num_angles)
+    return roi_inds, num_pixels, pixel_list
+
+
+def _process_rois(mesh, detector_size, num_rois):
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: wasn't there already a method with this name?

File: skxray/tests/test_diff_roi_choice.py [hunk 2]
```diff
@@ -185,3 +185,41 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
...
+
+    yy, xx = np.mgrid[:detector_size[0], :detector_size[1]]
+    y_ = (np.flipud(yy) - calibrated_center[1])
+    x_ = (xx - calibrated_center[0])
+    grid_values = np.float_(np.hypot(x_, y_))
+    angle_grid = np.arctan2(y_, x_)*180/(np.pi)
+
+    num_pixels_m = (np.bincount(ty.astype(int)))[1:]
+
+    return (ty.reshape(*detector_size))
```
Comment (tacaswell) [inline skxray/tests/test_diff_roi_choice.py]: Tests should not be returning anything

File: skxray/tests/test_diff_roi_choice.py [hunk 3]
```diff
@@ -185,3 +185,41 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
...
+
+
+
+if __name__=="__main__":
+    import matplotlib.pyplot as plt
```
Comment (tacaswell) [inline skxray/tests/test_diff_roi_choice.py]: and the main on this should run nose on the file

File: skxray/tests/test_diff_roi_choice.py (hunk 2 above)
Comment (sameera2004) [inline skxray/tests/test_diff_roi_choice.py]: Yes, this PR is in Progress

File: skxray/tests/test_diff_roi_choice.py (hunk 3 above)
Comment (sameera2004) [inline skxray/tests/test_diff_roi_choice.py]: Sure, I use matplotlib to see, whether code is working

File: skxray/diff_roi_choice.py (hunk 1 above)
Comment (sameera2004) [inline skxray/diff_roi_choice.py]: Yes _process_rings() changed it, now it will only process rings edge values. 
This module will take care of all the modules final processing of finding pixel_list, number of pixels, indices

File: skxray/tests/test_diff_roi_choice.py [hunk 4]
```diff
@@ -185,3 +187,60 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
...
+                           7, 7, 7, 7, 4, 4, 4, 5, 5, 5, 5, 6, 6,
+                           7, 7, 7, 7, 7, 7, 4, 5, 5, 5, 5, 5, 5,
+                           6, 6, 6, 7, 7, 7, 7, 7, 5, 5, 5, 5, 5,
+                           5, 6, 6, 6, 6, 7, 7, 7, 5, 5, 5, 5, 5,
+                           6, 6, 6, 6, 6, 7, 7, 7, 5, 5, 5, 5, 6,
+                           6, 6, 6, 6, 6, 7, 5, 5, 6, 6, 6, 6, 6,
+                           6])
+
+    assert_array_equal(num_pixels_m, num_pixels)
+    assert_array_equal(roi_inds, roi_inds_m)
```
Comment (sameera2004) [inline skxray/tests/test_diff_roi_choice.py]: divide_ring

Comment (sameera2004) [inline skxray/tests/test_diff_roi_choice.py]: @tacaswell  now this PR is ready for review

File: skxray/diff_roi_choice.py [hunk 5]
```diff
@@ -157,12 +160,12 @@ def roi_rings(img_dim, calibrated_center, num_rings,
 
     Returns
     -------
-    ring_vals : ndarray
-        edge values of the required  rings
-
     ring_inds : ndarray
         indices of the required rings
 
+    ring_vals : ndarray
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: Good catch

File: skxray/diff_roi_choice.py [hunk 6]
```diff
@@ -319,28 +322,52 @@ def _grid_values(img_dim, calibrated_center):
     return grid_values
 
 
-def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
+def _process_rings(num_rings, ring_vals):
     """
-    This will find the indices of the required rings, find the bin
-    edges of the rings, and count the number of pixels in each ring,
-    and pixels list for the required rings.
+    This will process the edge values of the each roi ring
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: Can you explain what sort of processing will be done by this function?

File: skxray/diff_roi_choice.py [hunk 7]
```diff
@@ -319,28 +322,52 @@ def _grid_values(img_dim, calibrated_center):
...
+    -------
     ring_vals : ndarray
         edge values of each ring
+        shape is (num_rings, 2)
 
-    ring_inds : ndarray
-        indices of the required rings
-        shape is ([detector_size[0]*detector_size[1]], )
+    """
+    ring_vals = np.array(ring_vals)
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: This can be done in one line as `np.asarray(ring_vas).reshape(-1, 2)`

File: skxray/diff_roi_choice.py [hunk 8]
```diff
@@ -360,18 +383,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
...
+
+def _process_rois(mesh, detector_size, num_rois):
+    """
+    This will find the indices of the required rois,
+    and count the number of pixels in each rois,
+    and pixels list for the required rois.
+
+    Parameters
+    ----------
+    mesh : ndarry
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: What is 'mesh'?

File: skxray/diff_roi_choice.py [hunk 9]
```diff
@@ -291,9 +293,10 @@ def roi_rings_step(img_dim, calibrated_center, num_rings,
     indx = ring_inds > 0
     ring_inds[indx] = (ring_inds[indx] + 1) // 2
 
-    (ring_inds, ring_vals, num_pixels,
-     pixel_list) = _process_rings(num_rings, img_dim,
-                                  ring_vals, ring_inds)
+    ring_vals = _process_rings(num_rings, np.asarray(ring_vals))
+
+    ring_inds, num_pixels, pixel_list = _process_rois(np.ravel(ring_inds),
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: This looks like you are un-raveling it so that you can then ravel it.

File: skxray/diff_roi_choice.py [hunk 10]
```diff
@@ -360,18 +383,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
         pixel indices for the required rings
 
     """
-    # find the pixel list
-    w = np.where(ring_inds > 0)
-    grid = np.indices((img_dim[0], img_dim[1]))
-    pixel_list = (grid[0]*img_dim[1] + grid[1]).flatten()[w]
+    mesh = np.zeros(detector_size, dtype=np.int64)
+    angles = np.linspace(0, 360, num_angles)
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: Can we work in radians internally?

File: skxray/diff_roi_choice.py [hunk 11]
```diff
@@ -360,18 +383,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
...
+    y_ = (np.flipud(yy) - calibrated_center[1])
+    x_ = (xx - calibrated_center[0])
+    grid_values = np.float_(np.hypot(x_, y_))
+    angle_grid = np.arctan2(y_, x_)*180/(np.pi)
 
-    ring_vals = np.array(ring_vals)
-    ring_vals = ring_vals.reshape(num_rings, 2)
+    angle_grid[grid_values > radius] = np.nan
+
+    for i in range(num_angles-1):
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: I think if you do this an `np.floor(angle_grid / (2*np.pi / num_angles)))` you can then use `bin_count` again.

File: skxray/diff_roi_choice.py [hunk 12]
```diff
@@ -319,28 +322,52 @@ def _grid_values(img_dim, calibrated_center):
...
-        indices of the required rings
-        shape is ([detector_size[0]*detector_size[1]], )
+    """
+    ring_vals = np.array(ring_vals)
+    ring_vals = ring_vals.reshape(num_rings, 2)
+
+    return ring_vals
+
+
+def roi_divide_circle(detector_size, radius, calibrated_center,
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: Should this also take a `start_angle` input?

File: skxray/diff_roi_choice.py (hunk 11 above)
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: and I would do the angle filtering _first_, then the radius filtering.

File: skxray/diff_roi_choice.py (hunk 12 above)
Comment (sameera2004) [inline skxray/diff_roi_choice.py]: No here I am using number of angles that ring(360) divided into

File: skxray/diff_roi_choice.py (hunk 11 above)
Comment (sameera2004) [inline skxray/diff_roi_choice.py]: Ok now I am thinking that will be better too,(angle filtering first, then the radius filtering)

File: skxray/diff_roi_choice.py (hunk 6 above)
Comment (sameera2004) [inline skxray/diff_roi_choice.py]: Added an explanation for _process_rings(num_rings, ring_vals):

**Comment by sameera2004** [comment]
Made changes according to comments. @tacaswell ready for review

File: skxray/diff_roi_choice.py (hunk 12 above)
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: but you can rotate where '0' is relative to the y-axis.

File: skxray/tests/test_diff_roi_choice.py (hunk 4 above)
Comment (tacaswell) [inline skxray/tests/test_diff_roi_choice.py]: What code generated this?

File: skxray/tests/test_diff_roi_choice.py [hunk 13]
```diff
@@ -185,3 +187,60 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
...
+                                              radius, calibrated_center,
+                                              num_angles)
+    ty = np.zeros(detector_size).ravel()
+    ty[pixel_list] = roi_inds
+
+    yy, xx = np.mgrid[:detector_size[0], :detector_size[1]]
+    y_ = (np.flipud(yy) - calibrated_center[1])
+    x_ = (xx - calibrated_center[0])
+    grid_values = np.float_(np.hypot(x_, y_))
+    angle_grid = np.arctan2(y_, x_)*180/(np.pi)
```
Comment (licode) [inline skxray/tests/test_diff_roi_choice.py]: use np.rad2deg

File: skxray/diff_roi_choice.py [hunk 14]
```diff
@@ -360,18 +383,85 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
...
+
+    Returns
+    -------
+    roi_inds : ndarray
+        indices of the required roi's
+        (after discarding zero values from the shape
+        ([detector_size[0]*detector_size[1]], )
+
+    num_pixels : ndarray
+        number of pixels in each roi's
```
Comment (licode) [inline skxray/diff_roi_choice.py]: This information can be calculated from roi_inds. Do you still want to return this again?

File: skxray/tests/test_diff_roi_choice.py [hunk 15]
```diff
@@ -185,3 +187,60 @@ def _helper_check(pixel_list, inds, num_pix, q_ring_val, calib_center,
                                                        - 0.000001)]]))[0][0]))
     assert_array_equal(zero_grid, data)
     assert_array_equal(num_pix, num_pixels)
+
+
+def test_roi_divide_circle():
+    num_rois = 3
```
Comment (celiafish) [inline skxray/tests/test_diff_roi_choice.py]: Please add description of this function

**Comment by sameera2004** [comment]
rotate_the_indices_roi_circle_divide

to make angles measured from vertical-anti clock wise

**Comment by sameera2004** [comment]
Thanks for the comments. Made changes , now the angle can start measuring from vertical direction, added documentation to roi_rectangels, made some changes to tests, added a detail commit, tests passing. Ready for PR review.

File: skxray/diff_roi_choice.py [hunk 16]
```diff
@@ -56,14 +59,17 @@
 logger = logging.getLogger(__name__)
 
 
-def roi_rectangles(num_rois, roi_data, detector_size):
+def roi_rectangles(num_rois, roi_data, detector_size, mask=None):
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: mask is not used any where and is not documented.

File: skxray/diff_roi_choice.py [hunk 17]
```diff
@@ -56,14 +59,17 @@
...
     """
+    This module will provide pixel indices, number of pixels, indices
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: function not module

File: skxray/diff_roi_choice.py [hunk 18]
```diff
@@ -319,28 +325,55 @@ def _grid_values(img_dim, calibrated_center):
...
+
+    calibrated_center : tuple
+        defining the center of the image
+        (column value, row value) (mm)
+
+    num_angles : int
+        number of angles ring divide into
+        angles are measured from horizontal-anti clock wise
+
+    rotate : {'Y', 'N'}, optional
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: Wouldn't this be better to make this an angle and add a `direction` kwarg`  ('cw' or 'ccw').

File: skxray/diff_roi_choice.py [hunk 19]
```diff
@@ -360,18 +389,83 @@ def _process_rings(num_rings, img_dim, ring_vals, ring_inds):
         pixel indices for the required rings
 
     """
+    #mesh = np.zeros(detector_size, dtype=np.int64)
+    angles = np.linspace(0, 360, num_angles)
```
Comment (tacaswell) [inline skxray/diff_roi_choice.py]: We really should work in radians internally.

File: skxray/diff_roi_choice.py (hunk 18 above)
Comment (sameera2004) [inline skxray/diff_roi_choice.py]: If we keep angle as a user input I don't think it will be a good idea. Because I want to divide the circle into same size wedges and later use that different wedges roi information(pixel indices, indices..) to run one time correlation and see whether there is a difference between those wedges correlation plots.

I kept the rotation as a optional input

File: skxray/diff_roi_choice.py (hunk 16 above)
Comment (sameera2004) [inline skxray/diff_roi_choice.py]: Sure, I will take out the mask for now, later I need to add it.

**Comment by tacaswell** [comment]
I thought you had changed the API for there to return the labeled arrays?

**Comment by sameera2004** [comment]
I changed that in the one time Correlation, this one has to rebase once that code merged. Therefore, have to review #254 first

**Comment by sameera2004** [comment]
Closed due to merge conflict (Github message : We can’t automatically merge this pull request.)

**Comment by tacaswell** [comment]
@sameera2004 You should just re-base on to current master, fix the merge confilcts, and then force-push back to this branch.
